### DISCUSSION TITLE
Possible to transform Geometry to Immutable?

### DISCUSSION THREAD

**Discussion by dbriard** [AvaloniaUI/Avalonia#11521] [Q&A] [answered]
Hi, I am facing to a weird issue... 

I copy/pasted the Shape classe to replace render with custom draw to add "effects"... 
Here the working Render method:
![image](https://github.com/AvaloniaUI/Avalonia/assets/3595443/14728cb6-d940-44b0-ad55-c5bd237fcca1)

Then I noticed that I forgot to remove the initial DrawGeometry call (highlighted above), so I removed it, and now I got exception here :
![image](https://github.com/AvaloniaUI/Avalonia/assets/3595443/25ff885b-90f8-492c-9c67-6f82d0111fb6)

I suppose it require the geometry to be immutable or frozen, but:

1) I don't understand why I do not get the exception when the DrawGeometry call is not removed (it can be before or after the custom).
2) As there is no Freeze() in Avalonia and I do not found a ToImmutable() on geometry, is there a way to freeze it?
3) Is there a better way than that to get SKPath from Geometry?
![image](https://github.com/AvaloniaUI/Avalonia/assets/3595443/f776c609-0257-4f87-abbb-a37ac92c8bce)

Thanks!

**Comment by Gillibald** [comment]
When the platform implementation is created properties of the mutable geometry are read. This is only allowed on the ui thread. The custom draw op is executed on the render thread.

**Reply by Gillibald** [reply, marked as answer]
So you should already pass an IGeometryImpl to your custom draw op when the control's render is called

**Reply by dbriard** [reply]
> So you should already pass an IGeometryImpl to your custom draw op when the control's render is called

Thanks, I assigned impl to a field in constructor now as I also need some ContainsFill method of the Geometry for HitTesting.
```
public ShapeCustomDrawOperation(Geometry geometry, ...)
            {
                Geometry = geometry;
                _impl = geometry.PlatformImpl;
                _size = geometry.Bounds.Size;
            }
```
